src/misc/support.cpp

```diff
@@ -289,6 +289,11 @@ static const std::deque<std_fs::path> &GetResourceParentPaths()
 			add_if_exists(resolved_dir / CANONICAL_PROJECT_NAME);
 		}
 	}
+
+	// Fifth priority is a best-effort fallback for --prefix installations
+	// into paths not pointed to by the system's XDG_DATA_ variables
+	add_if_exists(GetExecutablePath() / "../share" / CANONICAL_PROJECT_NAME);
```

Comment:
If this is only applicable w.r.t. --prefix, maybe instead this should use `add_if_exists(DATADIR / CANONICAL_PROJECT_NAME)` and pass `'-DDATADIR=@0@'.format(get_option('prefix') / get_option('datadir'))` as a compile flag.

This avoids a useless lookup for the executable path when it's not really relevant, and solves the problem where if one configures with `--prefix=/opt/dosbox --bindir=bin --datadir=resources` (or even worse, `--datadir=/opt/dosbox/idk/why/the/weird/path/resources`) it won't correctly calculate the datadir that the user configured with.

Reply:
It's a form of pure-relative lookup that will endure any `--prefix` and broken XDG variables.

Because it's relative against the executable, it tolerates the install tree being "carried around" into different locations, which is not-so-uncommon use-case among DOSBox users: GoG does it this way (each game directory has a copy of DOSBox sitting in the root), and so user often do the same with their own non-GoG games.

I don't think above suggestion handles this, correct?  (also: if the user specifies separate absolute paths for --prefix (say, `/opt`) and -datadir  (say, `/var/share`), the appending: `format(get_option('prefix') / get_option('datadir'))` will produce `/opt/var/share`.

I realize that some Linux programs carry configure-time directories into the source code and hard-code them as paths, but I've had enough pitfalls with thjis that I always prefer local and relative lookup.

Real-life  example, in college on a large multi-user machine I had `./configure`d a lot of free software into my own `--prefix$HOME/local/` area, but after backing up and migrating to other department systems (with a different top-level $HOME path and username pattern by the admins), some of the software stopped working citing lookup errors into my old $HOME directory - which is when I realized those packages hardcoded configure-time paths).

We have some mac users on old hardware, and in one ticket a user not able to setup a build environment and is asking another user for a zip of their install tree: provided the compiling user only mucks with `--prefix`, then they can zip and share without risk of breakage (and the receiving users doesn't have to bother with XDG paths).

Reply:
The only improvement I can see would be to handle non-absolute customized `--datadir` paths, that we can be certain exist relative to the executable:

`GetExecutablePath() / "../${non-absolute-datadir}" / CANONICAL_PROJECT_NAME)`

But I figured if someone's customizing `--datadir`, then they're probably a packager (or creating a Flatpak or AppImage) and know the connection to the XDG variables.

So for me, we're already in niche' work-around territory, and going even further is now adding more complexity into the code for a niche-of-a-niche case (so the scales are feeling tipped into over-engineering versus little benefit territory).

Reply:
> Because it's relative against the executable, it tolerates the install tree being "carried around" into different locations, which is a not-so-uncommon use-case among DOSBox users: GoG does it this way (each game directory has a copy of DOSBox sitting in the root), and so user often do the same with their own non-GoG games.

I thought that was why you check for the resources directory *next to* the executable... not why you check for an install-like structure where you find the parent directory of bin/ and then descend down into share.

> I don't think the above suggestion will handle this, correct? (also: if the user specifies separate absolute paths for --prefix (say, `/opt`) and -datadir (say, `/var/share`), the appending: `format(get_option('prefix') / get_option('datadir'))` will produce `/opt/var/share`.

Meson's path-join function and syntactic sugar operator does path joining, not string concatenation, so `join_paths('/usr', '/var/share')` or `'/usr' / '/var/share'` will always detect that /usr cannot be prepended to an absolute path, and outputs `/var/share`.

> I realize that some Linux programs carry configure-time directories into the source code and hard-code them as paths, but I've had enough pitfalls with thjis that I always prefer relative lookup.

The pitfall with relative lookup is that it will fail to find anything if the user has a non-default value for --datadir. You can compute the relativity of prefix vs. datadir at configure time and then embed that, I suppose. But you cannot assume it's `../share`.

Reply:
> The pitfall with relative lookup is that it will fail to find anything if the user has a non-default value for `--datadir`

Understood (and that's why it's a best-effort for --prefix only that will handle relocation of the install tree so long as datadir is default).

> Meson's path-join function and syntactic sugar operator does path joining, not string concatenation, so join_paths(`'/usr'`, `'/var/share'`) or `'/usr' / '/var/share'` will always detect that `/usr` cannot be prepended to an absolute path, and outputs `/var/share`.

Great! This helps a lot; will add your suggestion (early on as well), and we will carry this hard-coded combined path into the code - but in addition to the existing options, so we're not giving anything up. 

Meson's cross-platform path deductions at the build-system level transforms this from being tedious and hard in C/C++, to making this an easy one-line additional. Making the code easier to maintain, and giving users a better experience 🎉

Thank you, yet again, @eli-schwartz.

Reply:
(marking unresolved until implementated)

Reply:
Added! Thanks for the help, @eli-schwartz :+1: